File: Algorithms_4th_Edition/java/src/1/2/Rational.java
import java.util.Objects;

public class Rational {

    private final long numerator;
    private final long denominator;

    public Rational(long numerator, long denominator) {
        if(denominator == 0L)
            throw new IllegalArgumentException("denominator can not be 0");
        long l = gcd_sign(numerator, denominator);
        this.numerator = numerator / l;
        this.denominator = denominator / l;
    }

   public static long gcd(long p, long q)
   {
        if(q == 0) return p;
        long r = p % q;
        return gcd(q,r);
   }

   public static long gcd_sign(long p, long q)
   {
       long result = gcd(Math.abs(p),Math.abs(q));
       if(q < 0L && p < 0L)
           return -result;
       else
           return result;
   }
    public Rational plus(Rational b)
    {
        long n,d;
        n = this.numerator * b.denominator + this.denominator * b.numerator;
        d = this.denominator * b.denominator;
        return new Rational(n,d);
    }

    public Rational minus(Rational b)
    {
        long n,d;
        n = this.numerator * b.denominator - this.denominator * b.numerator;
        d =this.denominator * b.denominator;
        return new Rational(n,d);
    }

    public Rational times(Rational b)
    {
        long n,d;
        n = this.numerator * b.numerator;
        d = this.denominator * b.denominator;
        return new Rational(n,d);
    }

    public Rational divides(Rational b)
    {
        long n,d;
        n = this.numerator * b.denominator;
        d = this.denominator * b.numerator;
        return new Rational(n,d);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Rational rational = (Rational) o;
        return numerator == rational.numerator && denominator == rational.denominator;
    }

    @Override
    public int hashCode() {
        return Objects.hash(numerator, denominator);
    }

    @Override
    public String toString() {
        return this.numerator + "/" + this.denominator;
    }
}

File: Algorithms_4th_Edition/c/1/2/Rational.c
#include <stdio.h>
#include <stdlib.h>
#include "Rational.h"

long gcd_sign(long p, long q);

long gcd(long p,long q);

void initRational(Rational * a,int numerator, int denominator)
{
    if(denominator == 0)
    {
        fprintf(stderr,"the denominator : %d not zero",denominator);
        exit(EXIT_FAILURE);
    }
    long gcd_num;
    gcd_num = gcd_sign(numerator,denominator);
    a->numerator = numerator / gcd_num;
    a->denominator = denominator / gcd_num;
}

Rational * plus(Rational * r, const Rational * a, const Rational * b)
{
    long gcd_num,numerator,denominator;

    numerator = a->numerator * b->denominator + b->numerator * a->denominator;
    denominator = a->denominator * b->denominator;

    gcd_num = gcd_sign(numerator,denominator);
    r->numerator = numerator / gcd_num;
    r->denominator = denominator / gcd_num;
    return r;
}

Rational * minus(Rational * r, const Rational * a, const Rational * b)
{
    long gcd_num,numerator,denominator;

    numerator = a->numerator * b->denominator - b->numerator * a->denominator;
    denominator = a->denominator * b->denominator;

    gcd_num = gcd_sign(numerator,denominator);
    r->numerator = numerator / gcd_num;
    r->denominator = denominator / gcd_num;
    return r;
}

Rational * times(Rational * r, const Rational * a, const Rational * b)
{
    long gcd_num,numerator,denominator;

    numerator = a->numerator * b->numerator;
    denominator = a->denominator * b->denominator;

    gcd_num = gcd_sign(numerator,denominator);
    r->numerator = numerator / gcd_num;
    r->denominator = denominator / gcd_num;
    return r;
}

Rational * divides(Rational * r, const Rational * a, const Rational * b)
{
    long gcd_num,numerator,denominator;

    numerator = a->numerator * b->denominator;
    denominator = a->denominator * b->numerator;

    gcd_num = gcd_sign(numerator,denominator);
    r->numerator = numerator / gcd_num;
    r->denominator = denominator / gcd_num;
    return r;
}

int rationalEquals(const Rational * a,const Rational * b)
{
    return ( (a->numerator / b->numerator) / (a->denominator / b->denominator) ) == 1;
}

char * rationalToString(const Rational * a, char * str, size_t n)
{
    snprintf(str,n,"%ld/%ld",a->numerator,a->denominator);
    return str;
}

long gcd(long p,long q)
{
    if(q == 0) return p;
    long r = p % q;
    return gcd(q,r);
}

long gcd_sign(long p, long q)
{
    long result;
    result = gcd(p > 0 ? p : -p, q > 0 ? q : -q);
    if( p < 0 && q < 0)
        return -result;
    else
        return result;
}
